src/access/pg_tdetoast.c

```diff
@@ -793,3 +806,49 @@ pg_tde_fetch_toast_slice(Relation toastrel, Oid valueid, int32 attrsize,
 	systable_endscan_ordered(toastscan);
 	toast_close_indexes(toastidxs, num_indexes, AccessShareLock);
 }
+
+/* pg_tde extension */
+static void
+pg_tde_toast_tuple_externalize(ToastTupleContext *ttc, int attribute, int options)
+{
+	Datum	   *value = &ttc->ttc_values[attribute];
+	int32		data_size =0;
+	Pointer		dval = DatumGetPointer(*value);
+	char*		data_p;
+	char*		encrypted_data;
+    Relation    toastrel;
+    RelKeysData *keys;
+
+    /* Problem here is we want to encrypt the chunk using the toast relation key
+     * that is treated as a seperate relation. So we need to get the encryption keys
+     * for the toast table. Which is a performance penalty.
+     * TODO: come up with a better solution
+     */
+    toastrel = table_open(ttc->ttc_rel->rd_rel->reltoastrelid, NoLock);
+    keys = GetRelationKeys(toastrel->rd_locator);
+    table_close(toastrel, NoLock);
+    
+    if (VARATT_IS_SHORT(dval))
+	{
+		data_p = VARDATA_SHORT(dval);
+		data_size = VARSIZE_SHORT(dval) - VARHDRSZ_SHORT;
+	}
+	else if (VARATT_IS_COMPRESSED(dval))
+	{
+		data_p = VARDATA(dval);
+		data_size = VARSIZE(dval) - VARHDRSZ;
+	}
+	else
+	{
+		data_p = VARDATA(dval);
+		data_size = VARSIZE(dval) - VARHDRSZ;
+	}
```

Comment:
Both paths contain the same code. It could be:
```suggestion
	else
	{
		data_p = VARDATA(dval);
		data_size = VARSIZE(dval) - VARHDRSZ;
	}
```

Do I understand correctly that decompression happens later on? We don't have to care about it, right?

Reply:
Yes, that is correct. Fixed in latest version